### PR TITLE
Optimize separate

### PR DESCRIPTION
**Pull request by unnonouno**
I made a Separate function node. Originally `separate` function is implemented with `split_axis` and `reshape`. It creates a lot of `Reshape` function node. This PR reduces the number of function nodes.

### CODE REVIEW

File: chainer/functions/array/separate.py [hunk 1]
```diff
@@ -1,5 +1,38 @@
...
+        x_type = in_types[0]
+        if self.axis >= 0:
+            type_check.expect(self.axis < x_type.ndim)
+        else:
+            type_check.expect(-self.axis <= x_type.ndim)
+
+    def forward(self, inputs):
+        x, = inputs
+        self._xp = cuda.get_array_module(x)
+        xs = self._xp.split(x, x.shape[self.axis], self.axis)
```
Comment (okuta) [inline chainer/functions/array/separate.py]: How about following code?
```python
head  = (None,) * axis
ys = [x[head + (i,)] for i in six.moves.range(x.shape[self.axis])]
```

Comment (unnonouno) [reply chainer/functions/array/separate.py]: Is this faster even in cupy?

Comment (unnonouno) [reply chainer/functions/array/separate.py]: I tested the both with numpy but the performance is not different. I think `split` is simpler and better.

Comment (okuta) [reply chainer/functions/array/separate.py]: OK, I understand.

Comment (okuta) [reply chainer/functions/array/separate.py]: I got following result. My code is x10 faster when I use `numpy`.
```
In [38]: axis = 1
In [39]: xp = numpy
In [40]: x = xp.zeros((1000,1000))
In [41]: %timeit xs = xp.split(x, x.shape[axis], axis);ys = [xp.squeeze(y, axis) for y in xs]
100 loops, best of 3: 1.77 ms per loop

In [42]: %timeit head  = (None,) * axis; ys = [x[head + (i,)] for i in six.moves.range(x.shape[axis])]
1000 loops, best of 3: 253 µs per loop


In [46]: xp = cupy
In [47]: x = xp.zeros((1000,1000))

In [48]: %timeit xs = xp.split(x, x.shape[axis], axis);ys = [xp.squeeze(y, axis) for y in xs]
100 loops, best of 3: 2.68 ms per loop

In [50]: %timeit head  = (None,) * axis; ys = [x[head + (i,)] for i in six.moves.range(x.shape[axis])]
100 loops, best of 3: 3.05 ms per loop
```

Comment (unnonouno) [reply chainer/functions/array/separate.py]: We need to call `chainer.utils.force_array`. It makes the result worse.

Comment (okuta) [reply chainer/functions/array/separate.py]: Oh, Your right.

**Comment by okuta** [comment]
LGTM!

**Comment by delta2323** [comment]
Should we backport this PR?

**Pull request merged**